### ISSUE TITLE
No type parameters for method flatMap in for comprehension

### ISSUE THREAD

**Issue by jeroentervoorde** [scala/bug#10854]
Hi,

I am getting the following error in a for comprehension:

no type parameters for method flatMap
...
--- because ---
[error] argument expression's type is not compatible with formal parameter type;

It appears the have something to do with creating a new instance inside the for comprehension.
In one case the error can be worked around by making the return type of the for explicit. For the second case that also does not work.

I've made a reproducer here:

https://github.com/jeroentervoorde/type-inference-reproducer/blob/master/src/main/scala/Reproducer.scala

I tested this on 2.12.4, 2.12.5, 2.12.6, 2.13.0-M3 and dotty 0.8.0rc1
The errors only occur on on all versions except dotty.

I checked the bug database i found some bugs which look somewhat related but i am really not sure.

Thanks,
Jeroen 

Full error:

```
Reproducer.scala:36:7: no type parameters for method flatMap: (f: ((Main.wrappedRequest.Response, WrappedRequest[APIRequest])) => scala.concurrent.Future[S])(implicit executor: scala.concurrent.ExecutionContext)scala.concurrent.Future[S] exist so that it can be applied to arguments (((Main.wrappedRequest.Response, WrappedRequest[APIRequest])) => scala.concurrent.Future[WrappedResponse[newWrappedRequest.wrapped.Response]] forSome { val newWrappedRequest: WrappedRequest[APIRequest] })
[error]  --- because ---
[error] argument expression's type is not compatible with formal parameter type;
[error]  found   : ((Main.wrappedRequest.Response, WrappedRequest[APIRequest])) => scala.concurrent.Future[WrappedResponse[newWrappedRequest.wrapped.Response]] forSome { val newWrappedRequest: WrappedRequest[APIRequest] }
[error]     (which expands to)  ((WrappedResponse[Main.wrappedRequest.wrapped.Response], WrappedRequest[APIRequest])) => scala.concurrent.Future[WrappedResponse[newWrappedRequest.wrapped.Response]] forSome { val newWrappedRequest: WrappedRequest[APIRequest] }
[error]  required: ((Main.wrappedRequest.Response, WrappedRequest[APIRequest])) => scala.concurrent.Future[?S]
[error]     (which expands to)  ((WrappedResponse[Main.wrappedRequest.wrapped.Response], WrappedRequest[APIRequest])) => scala.concurrent.Future[?S]
[error]     _ <- wrappedApiClient.request(wrappedRequest)
[error]       ^
```

**Comment by joroKr21** [comment]
scala/scala#6421

I guess the problem appears because of path-dependent types and the way for comprehensions are desugared (the second `newWrappedRequest` is actually not the same one you defined - it has passed through a `map` function).

**Comment by jeroentervoorde** [comment]
Thanks. I'll confirm the fix once your PR's are merged. I guess the chance of them getting backported to 2.12 is slim?

**Comment by joroKr21** [comment]
I don't know if it will make in 2.13 either. There hasn't been any feedback so far. People are busy with the new collections.

**Comment by joroKr21** [comment]
Fixed by scala/scala#6421 in 2.13.0-M5
